### PR TITLE
Make helm charts consistent with how fields in spec are handled. (fleet-server only)

### PR DESCRIPTION
**Pull request by naemono**
related: #8192 
related: #8264
related: #8248

# Details of suggested change

This is the follow-up to the above PRs, where we are making the eck-fleet-server helm chart consistent in how it handles values that end up within `spec`, and making it backwards compatible.

### Implementation Note

1. The way this is implemented makes the previous behavior (having `.spec.field` in the values take precedence over just `field` in the values.
2. Mixing these 2 forms of values (some in spec in the values file, some not in spec) likely has unpredictable behavior. (for instance setting `spec.http.something`, and also setting `http.somethingElse`)

### CODE REVIEW

File: deploy/eck-stack/charts/eck-fleet-server/examples/fleet-server.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+version: 8.17.0-SNAPSHOT
+deployment:
+  replicas: 1
+  podTemplate:
+    spec:
+      serviceAccountName: fleet-server
```
Comment (pebrc) [inline deploy/eck-stack/charts/eck-fleet-server/examples/fleet-server.yaml:6]: There is a disconnect here between the name of this service account and the service account that is being generated. 

We are also setting the generated service account name as the `.spec.serviceAccountName` which is distinct from the service account in the pod spec. The latter is required for agent to work correctly the former is only relevant for the cross-namespace RBAC feature we have built into ECK ("Am I allowed to associated with an Elasticsearch in namespace x") . I am not sure if it is a problem to combine the two into one service account. Curious to get @barkbay 's perspective.

Comment (pebrc) [reply deploy/eck-stack/charts/eck-fleet-server/examples/fleet-server.yaml:6]: At the very least the service account names in this example have to be the same for the example to work.

File: deploy/eck-stack/charts/eck-fleet-server/examples/fleet-server.yaml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  namespace: default
+kibanaRef:
+  name: eck-kibana
+  namespace: default
+http:
+  service:
+    spec:
+      type: ClusterIP
+serviceAccount:
+  name: elastic-fleet-server
```
[the remark this answers is not in the document]
Comment (naemono) [reply deploy/eck-stack/charts/eck-fleet-server/examples/fleet-server.yaml:19]: image
All of the other charts expect `serviceAccount.name`, but render in the template `serviceAccountName`.

File: deploy/eck-stack/charts/eck-fleet-server/values.yaml [hunk 3]
```diff
@@ -28,46 +28,76 @@ labels: {}
...
+
+# policyID determines into which Agent Policy this Fleet Server will be enrolled.
+policyID: eck-fleet-server
+
+# The HTTP layer configuration for the Fleet Server Service.
+# ref: https://www.elastic.co/guide/en/cloud-on-k8s/current/k8s-elastic-agent-fleet-configuration.html#k8s-elastic-agent-fleet-configuration-customize-fleet-server-service
+#
+# http:
+
+# DaemonSet or Deployment specification for Fleet Server.
```
[the remark this answers is not in the document]
Comment (naemono) [reply deploy/eck-stack/charts/eck-fleet-server/values.yaml:77]: Very true. Let me fix this.

File: deploy/eck-stack/charts/eck-fleet-server/examples/fleet-server.yaml (hunk 1 above)
Comment (naemono) [reply deploy/eck-stack/charts/eck-fleet-server/examples/fleet-server.yaml:6]: I've updated this to make them consistent. What I tested was this: https://github.com/naemono/cloud-on-k8s/blob/helm-chart-image-fix-fleet/deploy/eck-stack/examples/agent/fleet-agents.yaml, which worked without issues, which only includes the sa name in the `podTemplate`.

File: deploy/eck-stack/charts/eck-fleet-server/examples/fleet-server.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (pebrc) [reply deploy/eck-stack/charts/eck-fleet-server/examples/fleet-server.yaml]:
```
find . -name 'values.yaml' -exec grep -H -A 1 '^# serviceAccoun' \{\} \;
./eck-stack/charts/eck-kibana/values.yaml:# serviceAccountName: ""
./eck-stack/charts/eck-kibana/values.yaml-
./eck-stack/charts/eck-elasticsearch/values.yaml:# serviceAccountName: ""
./eck-stack/charts/eck-elasticsearch/values.yaml-
./eck-stack/charts/eck-apm-server/values.yaml:# serviceAccountName: ""
./eck-stack/charts/eck-apm-server/values.yaml-
./eck-stack/charts/eck-logstash/values.yaml:# serviceAccountName: ""
./eck-stack/charts/eck-logstash/values.yaml-
./eck-stack/charts/eck-enterprise-search/values.yaml:# serviceAccountName: ""
./eck-stack/charts/eck-enterprise-search/values.yaml-
```

[the remark this answers is not in the document]
Comment (naemono) [reply deploy/eck-stack/charts/eck-fleet-server/examples/fleet-server.yaml]: Well, I'll make fleet consistent here, and then open an additional pr for beats/agent that's backwards compatible to make them also consistent.

[the remark this answers is not in the document]
Comment (pebrc) [reply deploy/eck-stack/charts/eck-fleet-server/examples/fleet-server.yaml]: I will try to take another look to understand better why we did that.

File: deploy/eck-stack/charts/eck-fleet-server/examples/fleet-server.yaml (hunk 1 above)
Comment (barkbay) [reply deploy/eck-stack/charts/eck-fleet-server/examples/fleet-server.yaml:6]: I'm not sure that's a problem either, that being said we made the choice to let the user specify a SA which can be different from the one used by the Pods (https://www.elastic.co/guide/en/cloud-on-k8s/master/k8s-restrict-cross-namespace-associations.html), and I'm not sure it is a reflected in the chart. [This comment](https://github.com/elastic/cloud-on-k8s/blob/f07d51dbe3a649923485511c2f18a421af2bd454/deploy/eck-stack/charts/eck-fleet-server/values.yaml#L104C1-L106) suggests that `serviceAccount.name` is used by the Pods while it is actually used by the [cross-namespace restriction mechanism](https://github.com/elastic/cloud-on-k8s/blob/f07d51dbe3a649923485511c2f18a421af2bd454/deploy/eck-stack/charts/eck-fleet-server/templates/fleet-server.yaml#L63)?

File: deploy/eck-stack/charts/eck-fleet-server/examples/fleet-server.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (pebrc) [reply deploy/eck-stack/charts/eck-fleet-server/examples/fleet-server.yaml]: After looking throught the git history I could not find a good argument why we structured `serviceAccount` property in the Agent and Beats charts in the way we did. 

In all the other charts `serviceAccountName` serves only the [cross-namespace RBAC feature](https://www.elastic.co/guide/en/cloud-on-k8s/current/k8s-restrict-cross-namespace-associations.html) but does not actually create the referenced service acount. 

In the Agent/Fleet Server and Beats chart  we actually need to generate a custom service account bound to a role with the use-case-dependent RBAC permissions for the application pods, so that the integrations the user wants to run work and can access the k8s API as needed. 

There are three **problems** with the approach taken: 

1. we coupled it with the cross-ns RBAC feature 
2. we allow to specify a namespace (does not makes sense as you pointed out) and undocumented also custom labels/annotations.
3. we have inconsistent approaches across eck-stack sub-charts

Given that we already released the all three affected charts I don't see how we can now stop supporting these attributes without potentially breaking customer installations.

1. the pointless namespace: its an ugly wart but does not hurt
2. the undocumented annotations/labels overrides. Let's keep them, they don't hurt either
3. the cross-ns coupled with the SA. Violates the principle of least privilege as users have to give the application pods an addtional get permission on Elasticsearch/Kibana  CRD to use the cross-ns RBAC feature that is not needed for the correct operation of the Beats/Agent pods. Again, the paramount concern here is backwards compatibility though. So I would keep it as is for now. Potentially we could introduce `serviceAccountName` in the future that would set the cross-ns SA if specified and otherwise use the `.serviceAccount.name` of the application pod. I don't like that solution very much as it would be very confusing even with extensive documentation. There is another wart that the SA that is being created by the Helm chart is not actually automatically used in the pod template, so the user has to be extra vigilent here and align all the ducks in one row.

**tl;dr**
My suggestion is as follows: 
* accept that we have a different approach with `.serviceAccount.name` for Beats, Agent and Fleet Server
* make sure that in the examples the names line up (you have already done that)
* create follow up issue to untangle the SA/cross-ns RBAC thing potentially in the future and document our decisions.

**Review by pebrc — APPROVED** [review]

**Comment by naemono** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|2.16|[](https://github.com/elastic/cloud-on-k8s/pull/8305)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)
